backend/mailru/mailru.go

```diff
@@ -0,0 +1,2356 @@
+	}
+	if apiErr, ok := err.(*api.FileErrorResponse); ok && apiErr.Status == 404 {
+		return "", fs.ErrorObjectNotFound
+	}
+	return "", err
+}
+
+// CleanUp permanently deletes all trashed files/folders
+func (f *Fs) CleanUp(ctx context.Context) error {
+	fs.Debugf(f, ">>> CleanUp")
+
+	token, err := f.accessToken()
+	if err != nil {
+		return err
+	}
+
+	data := url.Values{
+		"email":   {f.opt.Username},
+		"x-email": {f.opt.Username},
+	}
+	opts := rest.Opts{
+		Method: "POST",
+		Path:   "/api/m1/trashbin/empty",
+		Parameters: url.Values{
+			"access_token": {token},
+		},
+		Body:        strings.NewReader(data.Encode()),
+		ContentType: api.BinContentType,
+	}
+
+	var response api.CleanupResponse
+	err = f.pacer.Call(func() (bool, error) {
+		res, err := f.srv.CallJSON(&opts, nil, &response)
+		return shouldRetry(res, err, f, &opts)
+	})
+	if err != nil {
+		return err
+	}
+
+	switch response.StatusStr {
+	case "200":
+		return nil
+	default:
+		return fmt.Errorf("cleanup failed (%s)", response.StatusStr)
+	}
+}
+
+// About gets quota information
+func (f *Fs) About(ctx context.Context) (*fs.Usage, error) {
+	fs.Debugf(f, ">>> About")
+
+	token, err := f.accessToken()
+	if err != nil {
+		return nil, err
+	}
+	opts := rest.Opts{
+		Method: "GET",
+		Path:   "/api/m1/user",
+		Parameters: url.Values{
+			"access_token": {token},
+		},
+	}
+
+	var info api.UserInfoResponse
+	err = f.pacer.Call(func() (bool, error) {
+		res, err := f.srv.CallJSON(&opts, nil, &info)
+		return shouldRetry(res, err, f, &opts)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	total := info.Body.Cloud.Space.BytesTotal
+	used := int64(info.Body.Cloud.Space.BytesUsed)
+
+	usage := &fs.Usage{
+		Total: fs.NewUsageValue(total),
+		Used:  fs.NewUsageValue(used),
+		Free:  fs.NewUsageValue(total - used),
+	}
+	return usage, nil
+}
+
+// Put the object
+// Copy the reader in to the new object which is returned
+// The new object may have been created if an error is returned
+func (f *Fs) Put(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) (fs.Object, error) {
+	o := &Object{
+		fs:      f,
+		remote:  src.Remote(),
+		size:    src.Size(),
+		modTime: src.ModTime(ctx),
+	}
+	fs.Debugf(f, ">>> Put: %q %d '%v'", o.remote, o.size, o.modTime)
+	return o, o.Update(ctx, in, src, options...)
+}
+
+// Update an existing object
+// Copy the reader into the object updating modTime and size
+// The new object may have been created if an error is returned
+func (o *Object) Update(ctx context.Context, in io.Reader, src fs.ObjectInfo, options ...fs.OpenOption) error {
+	wrapIn := in
+	size := src.Size()
+	if size < 0 {
+		return errors.New("mailru does not support streaming uploads")
+	}
+
+	err := o.fs.mkParentDirs(ctx, o.absPath())
+	if err != nil {
+		return err
+	}
+
+	var (
+		fileBuf  []byte
+		fileHash []byte
+		newHash  []byte
+	)
+
+	// Check whether file is eligible for speedup method (put by hash)
```

Comment:
With mrhash in the hash framework you can just call `src.Hash(hash.Mrhash)` and if copying from the local backend, you'll get the hash which will be much quicker than spooling the file.

If you are copying from another backend (other than mailru or local) then you won't get the hash, and will need to spool it if you want it in advance.

If you want to see an example of another backend that does this see

https://github.com/rclone/rclone/blob/e2ce687f9377813039a1614658ef2d1952de7a90/backend/jottacloud/jottacloud.go#L1262-L1278